Userland/Applications/PixelPaint/FilterPreviewWidget.cpp

```diff
@@ -29,9 +34,22 @@ void FilterPreviewWidget::set_bitmap(RefPtr<Gfx::Bitmap> const& bitmap)
 
 void FilterPreviewWidget::set_filter(Filter* filter)
 {
-    if (filter)
+    if (filter) {
         filter->apply(*m_filtered_bitmap, *m_bitmap);
-    else
+
+        // If we have a layer set and the image has an active selection we only want the filter to apply to the
+        // selected region. This will walk the image and for every pixel that is outside the selection, restore it
+        // from the original bitmap.
+        if (m_layer && !m_layer->image().selection().is_empty()) {
+            for (int y = 0; y < m_filtered_bitmap->height(); ++y) {
+                for (int x = 0; x < m_filtered_bitmap->width(); ++x) {
+                    if (!m_layer->image().selection().is_selected(m_layer->location().translated(x, y))) {
+                        m_filtered_bitmap->set_pixel(x, y, m_bitmap->get_pixel(x, y));
+                    }
+                }
+            }
+        }
```

Comment:
hmmm, can't you blit here with some beforehand coordinate math? That would allow us to :yakspeed: with SIMD.

Reply:
If selected regions were strictly rectangular you could use blit, but with other selection tools being available now like wand and polygonal (and hopefully soon lasso), there is no guarantee that the selection is rectangular. So I don't think we can just blit with some coordinate math.

Reply:
okay then

Reply:
It's probably worth implementing a masked blit algo that does all the SIMD stuff, but that doesn't necessarily have to happen here in this PR or now.